### PR TITLE
chore(coverage): babel TypeScript & JS tests

### PR DESCRIPTION
**Pull request by xunilrj**
Instead of passing ```--language``` now we accept ```--suites```, which can pinpoint one specific test suite.
This PR also includes two more suites: typescript/babel and jsx/babel.

``` 
> cargo coverage --help
    Finished release [optimized] target(s) in 0.18s
     Running `target\release\xtask_coverage.exe --help`
cargo coverage
Run coverage command.
USAGE:
    cargo coverage <SUBCOMMAND> [option]
SUBCOMMANDS:
    compare             Compares output between two --json outputs
OPTIONS
    --markdown          Emits supported output into markdown format. Supported by `compare` subcommand.
    --json              Prints the test results in JSON. This mode will send all other test output and user messages to stderr.
    --detailed=[debug]  Prints a detailed summary at the end for all failing tests. Includes in depth details if set to `debug`.
    --suites=<IDS>      Runs the specified tests suites. Use comma as separator. Valid values are *, js, ts, jsx, js/262, ts/microsoft, ts/babel, jsx/babel. Default is "*".
    --filter=<file>     Filters out tests that don't match the query.
    --help              Prints this help.
```

### CODE REVIEW

File: xtask/coverage/src/jsx/jsx_babel.rs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+    fn run(&self) -> TestRunOutcome {
+        let jsx = Syntax::default().module();
+        let r = rslint_parser::parse(&self.code, 0, jsx);
+
+        let file = TestCaseFiles::single(self.name().to_string(), self.code.clone(), jsx);
+        if r.errors().is_empty() {
+            TestRunOutcome::Passed(file)
+        } else {
+            TestRunOutcome::IncorrectlyErrored {
```
Comment (MichaReiser) [inline xtask/coverage/src/jsx/jsx_babel.rs:41]: Are all these tests expected to pass? Are there no tests that should fail?

File: xtask/coverage/src/lib.rs [hunk 2]
```diff
@@ -147,16 +150,40 @@ pub fn run(
...
+    suites
+}
+
+pub fn check_file_encoding(path: &std::path::Path) -> Option<String> {
+    let buffer = std::fs::read(path).unwrap();
+    let bom = buffer.get(0..3);
+    //Utf16Be or // Utf16Le
+    if let Some(&[0xfe, 0xff, _] | &[0xff, 0xfe, _]) = bom {
+        None
+    } else {
```
Comment (MichaReiser) [inline xtask/coverage/src/lib.rs:184]: This method didn't work for me. I replaced it with a method that decodes utf16 files correctly instead in #2159

**Review by MichaReiser — COMMENTED** [review]
I like the `suites` change.

How did you verify the correctness of the test runner?

File: xtask/coverage/src/lib.rs (hunk 2 above)
Comment (xunilrj) [reply xtask/coverage/src/lib.rs:184]: I can rebase this when you merge it.

File: xtask/coverage/src/lib.rs [hunk 3]
```diff
@@ -148,16 +152,34 @@ pub fn run(
...
+    let mut ids: Vec<_> = suites.split(',').collect();
+
+    let mut suites: Vec<Box<dyn TestSuite>> = vec![];
+
+    while let Some(id) = ids.pop() {
+        match id {
+            "js" | "javascript" => ids.extend(["js/262"]),
+            "ts" | "typescript" => ids.extend(["ts/microsoft", "ts/babel"]),
+            "jsx" => ids.extend(["jsx/babel"]),
+            "*" => ids.extend(["js", "ts", "jsx"]),
```
Comment (ematipico) [inline xtask/coverage/src/lib.rs:166]: Can we have these ids inside some constants? It's difficult to understand what are the inputs of the CLI and the ids that actually matter.

**Review by ematipico — APPROVED** [review]

File: xtask/coverage/src/jsx/jsx_babel.rs (hunk 1 above)
Comment (xunilrj) [reply xtask/coverage/src/jsx/jsx_babel.rs:38]: > Are all these tests expected to pass? 
Yes

> Are there no tests that should fail?
Yes. But they are in a different folder. I was planning to include them later.

File: xtask/coverage/src/lib.rs (hunk 2 above)
Comment (xunilrj) [reply xtask/coverage/src/lib.rs]: Done.

File: xtask/coverage/src/ts/ts_babel.rs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+impl TestCase for BabelTypescriptTestCase {
+    fn name(&self) -> &str {
+        self.path
+            .components()
+            .rev()
+            .nth(1)
+            .and_then(|x| x.as_os_str().to_str())
+            .unwrap_or("")
```
[the remark this answers is not in the document]
Comment (xunilrj) [reply xtask/coverage/src/ts/ts_babel.rs]: Done.

File: xtask/coverage/src/jsx/jsx_babel.rs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+impl TestCase for BabelJsxTestCase {
+    fn name(&self) -> &str {
+        self.path
+            .components()
+            .rev()
+            .nth(1)
+            .and_then(|x| x.as_os_str().to_str())
+            .unwrap_or("")
```
[the remark this answers is not in the document]
Comment (xunilrj) [reply xtask/coverage/src/jsx/jsx_babel.rs]: Done.

File: xtask/coverage/src/ts/ts_babel.rs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    fn run(&self) -> TestRunOutcome {
+        let ts = Syntax::default().typescript();
```
[the remark this answers is not in the document]
Comment (xunilrj) [reply xtask/coverage/src/ts/ts_babel.rs]: This https://github.com/babel/babel/blob/main/packages/babel-parser/test/fixtures/typescript/options.json indicates that every typescript test is a module.

File: xtask/coverage/src/jsx/jsx_babel.rs [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    fn run(&self) -> TestRunOutcome {
+        let jsx = Syntax::default().module();
```
[the remark this answers is not in the document]
Comment (xunilrj) [reply xtask/coverage/src/jsx/jsx_babel.rs]: Now you got me!

The suite ```options.json``` does not say anything, and ```output.json```of each test says 

```
"type": "File"
```

So I think they are actually scripts.
I will change that.

[the remark this answers is not in the document]
Comment (xunilrj) [reply xtask/coverage/src/jsx/jsx_babel.rs]: Done.

**Comment by MichaReiser** [comment]
Can we add some checks for unknown nodes similar to #2171

File: xtask/coverage/src/lib.rs (hunk 3 above)
Comment (xunilrj) [reply xtask/coverage/src/lib.rs]: Is it better now?

**Comment by xunilrj** [comment]
> Can we add some checks for unknown nodes similar to #2171

Done.

File: xtask/coverage/src/jsx/jsx_babel.rs [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+            .find(|descendant| descendant.kind().is_unknown())
+        {
+            TestRunOutcome::IncorrectlyErrored {
+                files,
+                errors: vec![Diagnostic::new(
+                    0,
+                    Severity::Bug,
+                    "Unknown node in test that should pass",
+                )
+                .primary(unknown.text_range(), "")],
```
Comment (MichaReiser) [inline xtask/coverage/src/jsx/jsx_babel.rs:58]: We should use `text_trimmed_range` here.

File: xtask/coverage/src/jsx/jsx_babel.rs [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        );
+        let result = parse(&self.code, 0, source_type);
+
+        if let Some(unknown) = result
+            .syntax()
+            .descendants()
+            .find(|descendant| descendant.kind().is_unknown())
+        {
+            TestRunOutcome::IncorrectlyErrored {
+                files,
```
Comment (MichaReiser) [inline xtask/coverage/src/jsx/jsx_babel.rs:52]: Can we only emit this error if there are no other parsing errors. That nodes are `unknown` nodes if the parser detected parse errors makes sense and will be more useful when debugging.

File: xtask/coverage/src/lib.rs (hunk 3 above)
Comment (ematipico) [reply xtask/coverage/src/lib.rs]: I was actually thinking the other way around:

```rs
const TEST_262_SUITE: &str = "js/262";
```
but I think this is also fine :D

File: xtask/coverage/src/jsx/jsx_babel.rs (hunk 8 above)
Comment (xunilrj) [reply xtask/coverage/src/jsx/jsx_babel.rs:58]: Done.

File: xtask/coverage/src/jsx/jsx_babel.rs (hunk 9 above)
Comment (xunilrj) [reply xtask/coverage/src/jsx/jsx_babel.rs:52]: Done

**Pull request merged**